oauth.go

```diff
@@ -78,51 +76,19 @@ func getRefreshedToken(client *oidc.Client, t string) (jose.JWT, time.Time, erro
 	return token, identity.ExpiresAt, nil
 }
 
-// exchangeAuthenticationCode exchanges the authentication code with the oauth server for a access token
-func exchangeAuthenticationCode(client *oauth2.Client, code string) (oauth2.TokenResponse, error) {
-	return getToken(client, oauth2.GrantTypeAuthCode, code)
-}
-
-// getUserinfo is responsible for getting the userinfo from the IDPD
-func getUserinfo(client *oauth2.Client, endpoint string, token string) (jose.Claims, error) {
-	req, err := http.NewRequest(http.MethodGet, endpoint, nil)
-	if err != nil {
-		return nil, err
-	}
-	req.Header.Set(authorizationHeader, fmt.Sprintf("Bearer %s", token))
-
-	resp, err := client.HttpClient().Do(req)
-	if err != nil {
-		return nil, err
-	}
-	if resp.StatusCode != http.StatusOK {
-		return nil, errors.New("token not validate by userinfo endpoint")
-	}
-	content, err := ioutil.ReadAll(resp.Body)
-	if err != nil {
-		return nil, err
-	}
-	var claims jose.Claims
-	if err := json.Unmarshal(content, &claims); err != nil {
-		return nil, err
-	}
-
-	return claims, nil
-}
-
 // getToken retrieves a code from the provider, extracts and verified the token
-func getToken(client *oauth2.Client, grantType, code string) (oauth2.TokenResponse, error) {
+func getToken(config *oauth2.Config, grantType, code string) (*oauth2.Token, error) {
 	start := time.Now()
-	token, err := client.RequestToken(grantType, code)
+	token, err := config.Exchange(context.Background(), code)
```

Comment:
Would this only work for GrantTypeAuthCode now? ... given it's not taking the grant type into account .. https://github.com/golang/oauth2/blob/master/oauth2.go#L188-L197 .. i.e. the call https://github.com/gambol99/keycloak-proxy/pull/361/files#diff-b3c698f6274c18640399beb95db68658R63 will fail

Reply:
@gambol99 I can be wrong. But I believe that based on the oauth2 documentation, we no longer need to specify the grant type, because the API can automatically handle it. See: https://github.com/golang/oauth2/blob/master/oauth2.go#L81. 

We have this at line: https://github.com/gambol99/keycloak-proxy/pull/361/files#diff-31a4a7824225629df071a1cba8b785d5R72. And the password grant type I believe is covered here: https://github.com/gambol99/keycloak-proxy/pull/361/files#diff-ad348c27fa1bf8571baa68994a271da3R109.

Looking at the coverage tests seems to hit the calls to RefreshToken. Anyways, I would like to double-check and reproduce the failure you mentioned. Would you mind to provide which steps you've done?

Reply:
hi @abstractj .. yep, your right .. i rebased and checked, it does appear to refresh the token.

```shell
1.5275125002257237e+09	debug	keycloak-proxy/middleware.go:323	access permitted to resource	{"access": "permitted", "email": "gambol99@gmail.com", "expires": 4.774277085, "resource": "/*"}
1.5275125079456906e+09	info	keycloak-proxy/middleware.go:140	accces token for user has expired, attemping to refresh the token	{"client_ip": "127.0.0.1:58154", "email": "gambol99@gmail.com"}
1.527512508070129e+09	info	keycloak-proxy/middleware.go:176	injecting the refreshed access token cookie	{"client_ip": "127.0.0.1:58154", "cookie_name": "kc-access", "email": "gambol99@gmail.com", "expires_in": 59.929872189}
```

Reply:
@gambol99 cool, glad to hear. I changed the PR to be merged under the *develop* branch.